code/game/turfs/baseturfs.dm

```diff
@@ -0,0 +1,188 @@
+			if(!length(baseturfs))
+				baseturfs = list(baseturfs)
+			var/list/old_baseturfs = baseturfs.Copy()
+			if(!isclosedturf(src))
+				old_baseturfs += type
+			newT = ChangeTurf(fake_turf_type, null, flags)
+			newT.assemble_baseturfs(initial(fake_turf_type.baseturfs)) // The baseturfs list is created like roundstart
+			if(!length(newT.baseturfs))
+				newT.baseturfs = list(baseturfs)
+			// The old baseturfs are put underneath, and we sort out the unwanted ones
+			newT.baseturfs = baseturfs_string_list(old_baseturfs + (newT.baseturfs - GLOB.blacklisted_automated_baseturfs), newT)
+			return newT
+		if(!length(baseturfs))
+			baseturfs = list(baseturfs)
+		if(!isclosedturf(src))
+			new_baseturfs = list(type) + new_baseturfs
+		baseturfs = baseturfs_string_list(baseturfs + new_baseturfs, src)
+		return ChangeTurf(fake_turf_type, null, flags)
+	if(!length(baseturfs))
+		baseturfs = list(baseturfs)
+	if(!isclosedturf(src))
+		baseturfs = baseturfs_string_list(baseturfs + type, src)
+	var/turf/change_type
+	if(length(new_baseturfs))
+		change_type = new_baseturfs[new_baseturfs.len]
+		new_baseturfs.len--
+		if(new_baseturfs.len)
+			baseturfs = baseturfs_string_list(baseturfs + new_baseturfs, src)
+	else
+		change_type = new_baseturfs
+	return ChangeTurf(change_type, null, flags)
+
+// Copy an existing turf and put it on top
+// Returns the new turf
+/turf/proc/CopyOnTop(turf/copytarget, ignore_bottom=1, depth=INFINITY, copy_air = FALSE)
+	var/list/new_baseturfs = list()
+	new_baseturfs += baseturfs
+	new_baseturfs += type
+
+	if(depth)
+		var/list/target_baseturfs
+		if(length(copytarget.baseturfs))
+			// with default inputs this would be Copy(clamp(2, -INFINITY, baseturfs.len))
+			// Don't forget a lower index is lower in the baseturfs stack, the bottom is baseturfs[1]
+			target_baseturfs = copytarget.baseturfs.Copy(clamp(1 + ignore_bottom, 1 + copytarget.baseturfs.len - depth, copytarget.baseturfs.len))
+		else if(!ignore_bottom)
+			target_baseturfs = list(copytarget.baseturfs)
+		if(target_baseturfs)
+			target_baseturfs -= new_baseturfs & GLOB.blacklisted_automated_baseturfs
+			new_baseturfs += target_baseturfs
+
+	var/turf/newT = copytarget.copyTurf(src, copy_air)
+	newT.baseturfs = baseturfs_string_list(new_baseturfs, newT)
+	return newT
+
+/// Tries to find the given type in baseturfs.
+/// If found, returns how deep it is for use in other baseturf procs, or null if it cannot be found.
+/// For example, this number can be passed into ScrapeAway to scrape everything until that point.
+/turf/proc/depth_to_find_baseturf(baseturf_type)
+	var/index = baseturfs.Find(baseturf_type)
+	if (index == 0)
+		return null
+	return baseturfs.len - index + 1
+
+/// Returns the baseturf at the given depth.
+/// For example, baseturf_at_depth(1) will give the baseturf that would show up when scraping once.
+/turf/proc/baseturf_at_depth(index)
+	ASSERT(isnum(index))
```

Comment:
does this need to be runtime? don't like this

Reply:
As in should this be `TEST_ONLY_ASSERT`? Or should numbers be allowed in?

In which case my answers would be sure (though I would imagine a lot of baseturf stuff is shuttles, which might not be properly tracked over runtimes), and I don't see why

Reply:
I don't like that it asserts on live, cause this is a vaguely hot case especially for init, and I want to avoid it if possible. Seems superfulous 
numbers being the only valid input is sane.

Reply:
```suggestion
	TEST_ONLY_ASSERT(isnum(index))
```